### PR TITLE
Add register and login functionality

### LINKED ISSUE #54: Implement Server Side session management
use `next-auth`

### PR DESCRIPTION
**Pull request by dewa-himanshu24**
closes #54

### CODE REVIEW

File: pages/api/login.ts [hunk 1]
```diff
@@ -1,29 +1,51 @@
...
+import bcrypt from 'bcrypt'
+import jwt from 'jsonwebtoken'
 
 const prisma = new PrismaClient()
 
 type Data = {
   message: string,
-  xBhaktToken: any
+  xBhaktToken?: any
+  decoded?: any
```
Comment (jainendra) [inline pages/api/login.ts]: This is not required

File: pages/api/login.ts [hunk 2]
```diff
@@ -1,29 +1,51 @@
...
+            id: data.id
+          }
+
+          const xBhaktToken = jwt.sign(payload, secret, {expiresIn: '1s'});
+          console.log(jwt.verify(xBhaktToken,secret));
+
+          res.status(200).json({
+            message: "Successfully loggedin",
+            xBhaktToken: xBhaktToken,
+            decoded: jwt.verify(xBhaktToken,secret)
```
Comment (jainendra) [inline pages/api/login.ts]: This is not required

File: pages/api/login.ts [hunk 3]
```diff
@@ -1,29 +1,51 @@
...
+          }
+          const secret = "panditaijwttoken";
+
+          const payload = {
+            email: data.email,
+            id: data.id
+          }
+
+          const xBhaktToken = jwt.sign(payload, secret, {expiresIn: '1s'});
+          console.log(jwt.verify(xBhaktToken,secret));
```
Comment (jainendra) [inline pages/api/login.ts]: Lets use
`console.log(``User ${data.id} successfully logged in``)`

File: pages/api/login.ts [hunk 4]
```diff
@@ -1,29 +1,51 @@
...
-      }))
+    prisma.bhakt.findFirstOrThrow({ where: { email: email } })
+      .then(data =>
+        bcrypt.compare(password, data.password_hash).then((match) => {
+          if (!match) {
+            res.status(400).json({
+              message: "Wrong Credential!",
+            })
+          }
+          const secret = "panditaijwttoken";
```
Comment (jainendra) [inline pages/api/login.ts]: process.env.APP_AUTH_JWT_SECRET

File: pages/api/register.ts [hunk 5]
```diff
@@ -25,21 +26,24 @@ export default async function handler(
...
-      .then(data => res.status(200).json({
-        message: "Successfully registered"
-      }))
-      .catch((err) => {
-        console.log(err);
-        res.status(400).json({
-          message: "Something went wrong"
-        })
-      });
+    const hashed_password = bcrypt.hash(password, 10).then((hashed_password: string) => {
```
Comment (jainendra) [inline pages/api/register.ts]: move 10 in a variable `saltRounds`

File: pages/api/login.ts [hunk 6]
```diff
@@ -1,29 +1,49 @@
...
-        xBhaktToken: data.id
-      }))
+    prisma.bhakt.findFirstOrThrow({ where: { email: email } })
+      .then(data =>
+        bcrypt.compare(password, data.password_hash).then((match) => {
+          if (!match) {
+            res.status(400).json({
+              message: "Wrong Credential!",
+            })
+          }
```
Comment (jainendra) [inline pages/api/login.ts:29]: We have to stop execution of our code here using return or keep the rest of the code in else part

**Comment by jainendra** [comment]
@dewa-himanshu24 Please also update env.example file in this PR

**Comment by dewa-himanshu24** [comment]
Done

File: pages/api/login.ts [hunk 7]
```diff
@@ -1,29 +1,49 @@
...
-        message: "Successfully loggedin",
-        xBhaktToken: data.id
-      }))
+    prisma.bhakt.findFirstOrThrow({ where: { email: email } })
+      .then(data =>
+        bcrypt.compare(password, data.password_hash).then((match) => {
+          if (!match) {
+            res.status(400).json({
+              message: "Wrong Credential!",
+            })
```
Comment (jainendra) [inline pages/api/login.ts:27]: @dewa-himanshu24 I think you missed this, adding again
- We'll need to add either return statement or put the bottom part of the code in an else block in order to stop further execution

**Review by jainendra — APPROVED** [review]

**Pull request merged**